File: Python/0020_valid_parentheses/0020_valid_parentheses.py
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = Stack()
        map = {']': '[', ')': '(', '}': '{'}
        for i in range(0, len(s)):
            if (s[i] == '(') | (s[i] == '[') | (s[i] == '{'):
                stack.push(s[i])
            else:
                if stack.isEmpty():
                    return False
                top = stack.peek()
                if top != map[s[i]]:
                    return False
                stack.pop()
        return stack.isEmpty()

    
class Stack:
    def __init__(self):
        self.items = []

    def isEmpty(self):
        return len(self.items) == 0

    def push(self, item):
        self.items.append(item)

    def pop(self):
        return self.items.pop()

    def peek(self):
        if not self.isEmpty():
            return self.items[len(self.items) - 1]

    def size(self):
        return len(self.items)

File: C/0020_valid_parentheses/0020_valid_parentheses.c
void push(char *stack, int *num, char element) {
    stack[*num] = element;
    (*num)++; 
}

char top(char *stack, int *num) {
    return stack[*num - 1];
}

void pop(char *stack, int *num) {
    (*num)--; 
}


bool isValid(char *s) {
    char stack[10000];
    int num = 0;
    for (int i = 0; i < strlen(s); i++) {
        if (s[i] == '(' || s[i] == '[' || s[i] == '{') {
            push(stack, &num, s[i]);
        }
        else {
            if (num == 0) {
                return false;
            }
            char stackTop = top(stack, &num);
            if ((stackTop == '[' && s[i] == ']') ||
                (stackTop == '(' && s[i] == ')') ||
                (stackTop == '{' && s[i] == '}')) {
                pop(stack, &num);
            }
            else {
                return false;
            }
        }
    }
    return num == 0;
}